services/network/networksecuritygroup/wssd.go

```diff
@@ -0,0 +1,360 @@
+		wssdCloudNSG.Tags = tags.MapToProto(networkNSG.Tags)
+	}
+
+	if networkNSG.SecurityGroupPropertiesFormat != nil {
+		nsgRules, err := getWssdNetworkSecurityGroupRules(networkNSG.SecurityRules)
+		if err != nil {
+			return nil, err
+		}
+		defaultNsgRules, err := getWssdNetworkSecurityGroupRules(networkNSG.DefaultSecurityRules)
+		if err != nil {
+			return nil, err
+		}
+		wssdCloudNSG.Networksecuritygrouprules = append(nsgRules, defaultNsgRules...)
+	}
+
+	return wssdCloudNSG, nil
+}
+
+// getWssdNetworkSecurityGroupRules converts our internal representation of a networksecuritygroup rule (network.SecurityRule) to the cloud network security group rule protobuf used by wssdcloudagent (wssdnetwork.NetworkSecurityGroupRule)
+func getWssdNetworkSecurityGroupRules(securityRules *[]network.SecurityRule) (wssdNSGRules []*wssdcloudnetwork.NetworkSecurityGroupRule, err error) {
+	if securityRules == nil || len(*securityRules) <= 0 {
+		return
+	}
+
+	for _, rule := range *securityRules {
+		if rule.SecurityRulePropertiesFormat == nil {
+			continue
+		}
+
+		wssdCloudNSGRule := &wssdcloudnetwork.NetworkSecurityGroupRule{}
+
+		if rule.Name == nil {
+			return nil, errors.Wrapf(errors.InvalidInput, "Network Security Rule name not specified")
+		}
+		wssdCloudNSGRule.Name = *rule.Name
+
+		if rule.Description != nil {
+			wssdCloudNSGRule.Description = *rule.Description
+		}
+
+		if strings.EqualFold(string(rule.Protocol), string(network.SecurityRuleProtocolAsterisk)) {
+			wssdCloudNSGRule.Protocol = wssdcloudcommon.Protocol_All
+		} else if strings.EqualFold(string(rule.Protocol), string(network.SecurityRuleProtocolTCP)) {
+			wssdCloudNSGRule.Protocol = wssdcloudcommon.Protocol_Tcp
+		} else if strings.EqualFold(string(rule.Protocol), string(network.SecurityRuleProtocolUDP)) {
+			wssdCloudNSGRule.Protocol = wssdcloudcommon.Protocol_Udp
+		} else {
+			return nil, errors.Wrapf(errors.InvalidInput, "Unknown Protocol %s specified", rule.Protocol)
+		}
+
+		if rule.SourceAddressPrefix != nil {
+			wssdCloudNSGRule.SourceAddressPrefix = *rule.SourceAddressPrefix
+		} else if rule.SourceAddressPrefixes != nil {
+			concatRule := ""
+			for _, prefix := range *rule.SourceAddressPrefixes {
+				concatRule += prefix
+			}
+			wssdCloudNSGRule.SourceAddressPrefix = concatRule
+		}
+
+		if rule.DestinationAddressPrefix != nil {
+			wssdCloudNSGRule.DestinationAddressPrefix = *rule.DestinationAddressPrefix
+		} else if rule.DestinationAddressPrefixes != nil {
+			concatRule := ""
+			for _, prefix := range *rule.DestinationAddressPrefixes {
+				concatRule += prefix
+			}
+			wssdCloudNSGRule.DestinationAddressPrefix = concatRule
+		}
+
+		if rule.SourcePortRange != nil {
+			wssdCloudNSGRule.SourcePortRange = *rule.SourcePortRange
+		} else if rule.SourcePortRanges != nil {
+			concatRule := ""
+			for _, prefix := range *rule.SourcePortRanges {
+				concatRule += prefix
+			}
+			wssdCloudNSGRule.SourcePortRange = concatRule
+		}
```

Comment:
Should we parse any malformed inputs ?

Reply:
I think the logic for it is complicated enough that I'd want to just let NC handle it, but you're right, we could add in the future.